Cluster mempool implementation

sdaftuar opened the pull request:
This is a draft implementation of the cluster mempool design described in #27677.  I'm opening this as a draft PR now to share the branch I'm working on with others, so that we can start to think about in-progress projects (like package relay, package validation, and package rbf) in the context of this design.  Also, I can use some help from others for parts of this work, including the interaction between the mempool and the wallet, and also reworking some of our existing test cases to fit a cluster-mempool world.

Note that the design of this implementation is subject to change as I continue to iterate on the code (to make the code more hygienic and robust, in particular). At this point though I think the performance is pretty reasonable and I'm not currently aware of any bugs.  There are some microbenchmarks added here, and some improved fuzz tests; it would be great if others ran both of those on their own hardware as well and reported back on any findings.

This branch implements the following observable behavior changes:

- Maintains a partitioning of the mempool into connected clusters
- Each cluster is sorted ("linearized") either using an optimal sort, or an ancestor-feerate-based one, depending on the size of the cluster (thanks to @sipa for this logic)
- Transaction selection for mining is updated to use the cluster linearizations
- Mempool eviction is updated to use the cluster linearizations
- The RBF rules are updated to drop the requirement that no new inputs are introduced, and to change the feerate requirement to instead check that the mining score of a replacement transaction exceed the mining score of the conflicted transactions
- The CPFP carveout rule is eliminated (it doesn't make sense in a cluster-limited mempool)
- The ancestor and descendant limits are no longer enforced.
- New cluster count/cluster vsize limits are now enforced instead.

Some less observable behavior changes:
- The cached ancestor and descendant data are dropped from the mempool, along with the multi_index indices that were maintained to sort the mempool by ancestor and descendant feerates. For compatibility (eg with wallet behavior or RPCs exposing this), this information is now calculated dynamically instead.
- The ancestor and descendant walking algorithms are now implemented using `epoch`s (resulting in a significant performance improvement, according to the benchmarks I've looked at)

Still to do:

- [x] More comparisons between this branch and master on historical data to compare validation speed (accepting loose transactions, processing RBF transactions, validating a block/postprocessing, updating the mempool for a reorg).
- [x] More historical data analysis to try to evaluate the likely impact of setting the cluster size limits to varying values (to motivate what values we should ultimately pick). [DONE, see [this post](https://delvingbitcoin.org/t/research-into-the-effects-of-a-cluster-size-limited-mempool-in-2023/794/1)]
- [ ] Updating wallet code to be cluster-aware (including mini_miner and coin selection)
- [ ] Rework many of our functional tests to be cluster-aware
- [ ] Figure out what package validation and package RBF rules should be in this design
- [x] Rework the `partially_downloaded_block` fuzz target to not add duplicate transactions to the mempool (#29990).
- [x] Update RBF logic to ensure that replacements always strictly improve the mempool.
- [ ] Figure out how we want to document our RBF policy (preserve historical references to BIP 125 or previous Bitcoin Core behaviors vs clean slate documentation?)

For discussion/feedback:

 - How significant is it to be dropping the CPFP carveout rule?  Does that affect how we will ultimately want to stage new mempool deployment?
 - How well do the proposed RBF rules meet everyone's use cases?
 - What design improvements can we make to the cluster tracking implementation?
 - The ZMQ callbacks that occur when a block is found will happen in a slightly different order, because we now will fully remove all transactions occurring in a block from the mempool before removing any conflicts. Is this a problem?

sdaftuar committed:
Expose cluster information via rpc

Co-authored-by: glozow <gloriajzhao@gmail.com>

sdaftuar committed:
Limit mempool size based on chunk feerate

Rather than evicting the transactions with the lowest descendant feerate,
instead evict transactions that have the lowest chunk feerate.

Once mining is implemented based on choosing transactions with highest chunk
feerate (see next commit), mining and eviction will be opposites, so that we
will evict the transactions that would be mined last.

sdaftuar committed:
Select transactions for blocks based on chunk feerate

sdaftuar committed:
Add new (unused) limits for cluster size/count

sdaftuar committed:
Do not allow mempool clusters to exceed configured limits

Include an adjustment to mempool_tests.cpp due to the additional memory used by
txgraph.

Includes a temporary change to the mempool_ephemeral_dust.py functional test,
due to validation checks being reordered. This change will revert once the RBF
rules are changed in a later commit.

sdaftuar committed:
Implement new RBF logic for cluster mempool

With a total ordering on mempool transactions, we are now able to calculate a
transaction's mining score at all times. Use this to improve the RBF logic:

- we no longer enforce a "no new unconfirmed parents" rule

- we now require that the mempool's feerate diagram must improve in order
  to accept a replacement

- the topology restrictions for conflicts in the package rbf setting have been
  eliminated

Revert the temporary change to mempool_ephemeral_dust.py that were previously
made due to RBF validation checks being reordered.

Co-authored-by: Gregory Sanders <gsanders87@gmail.com>, glozow <gloriajzhao@gmail.com>

sdaftuar committed:
Remove the ancestor and descendant indices from the mempool

sdaftuar committed:
Use cluster linearization for transaction relay sort order

Previously, transaction batches were first sorted by ancestor count and then
feerate, to ensure transactions are announced in a topologically valid order,
while prioritizing higher feerate transactions. Ancestor count is a crude
topological sort criteria, so replace this with linearization order so that the
highest feerate transactions (as would be observed by the mining algorithm) are
relayed before lower feerate ones, in a topologically valid way.

This also fixes a test that only worked due to the ancestor-count-based sort
order.

sdaftuar committed:
Remove CTxMemPool::GetSortedDepthAndScore

The mempool clusters and linearization permit sorting the mempool topologically
without making use of ancestor counts (as long as the graph is not oversized).

Co-authored-by: Pieter Wuille <pieter@wuille.net>

sdaftuar committed:
Reimplement GetTransactionAncestry() to not rely on cached data

In preparation for removing ancestor data from CTxMemPoolEntry, recalculate the
ancestor statistics on demand wherever needed.

sdaftuar committed:
rpc: Calculate ancestor data from scratch for mempool rpc calls

sdaftuar committed:
Remove dependency on cached ancestor data in mini-miner

sdaftuar committed:
Stop enforcing ancestor size/count limits

The cluster limits should be sufficient.

Co-Authored-By: Gregory Sanders <gsanders87@gmail.com>

sdaftuar committed:
policy: Remove CPFP carveout rule

The addition of a cluster size limit makes the CPFP carveout rule useless,
because carveout cannot be used to bypass the cluster size limit. Remove this
policy rule and update tests to no longer rely on the behavior.

sdaftuar committed:
Use mempool/txgraph to determine if a tx has descendants

Remove a reference to GetCountWithDescendants() in preparation for removing
this function and the associated cached state from the mempool.

sdaftuar committed:
Calculate descendant information for mempool RPC output on-the-fly

This is in preparation for removing the cached descendant state from the
mempool.

sdaftuar committed:
Stop enforcing descendant size/count limits

Cluster size limits should be enough.

sdaftuar committed:
wallet: Replace max descendantsize with clustersize

With the descendant size limits removed, replace the concept of "max number of
descendants of any ancestor of a given tx" with the cluster count of the cluster
that the transaction belongs to.

sdaftuar committed:
mempool: Remove unused function CalculateDescendantMaximum

sdaftuar committed:
mempool: eliminate accessors to mempool entry ancestor/descendant cached state

sdaftuar committed:
Remove unused members from CTxMemPoolEntry

sdaftuar committed:
Remove mempool logic designed to maintain ancestor/descendant state

sdaftuar committed:
mempool: addUnchecked no longer needs ancestors

sdaftuar committed:
bench: add more mempool benchmarks

Add benchmarks for:

  - mempool update time when blocks are found
  - adding a transaction
  - performing the mempool's RBF calculation
  - calculating mempool ancestors/descendants

sdaftuar committed:
fuzz: try to add more code coverage for mempool fuzzing

Including test coverage for mempool eviction and expiry

sdaftuar committed:
Remove unused limits from CalculateMemPoolAncestors

sdaftuar committed:
Simplify ancestor calculation functions

Now that ancestor calculation never fails (due to ancestor/descendant limits
being eliminated), we can eliminate the error handling from
CalculateMemPoolAncestors.

sdaftuar committed:
Make removeConflicts private

sdaftuar committed:
==== END CLUSTER IMPLEMENTATION ====

sdaftuar committed:
==== BEGIN MEMPOOL CLEANUP ====

sdaftuar committed:
==== END MEMPOOL CLEANUP ====

sdaftuar committed:
==== BEGIN OPTIMIZATIONS ====

sdaftuar committed:
test: remove rbf carveout test from mempool_limit.py

sdaftuar commented on src/bench/mempool_stress.cpp:
```diff
@@ -83,26 +86,104 @@ static std::vector<CTransactionRef> CreateOrderedCoins(FastRandomContext& det_ra
+    CTxMemPool& pool = *testing_setup.get()->m_node.mempool;
+
+    std::vector<CTransactionRef> transactions;
+    // Create 1000 clusters of 100 transactions each
```
The immediate downside to a cap on number of clusters is that singleton, high-feerate transactions would not be accepted.  And I don't think we need to -- the only places where having more clusters makes us slower is in eviction and mining, and for both of those use cases we could improve performance (if we need to) by maintaining the relevant heap data structures (or something equivalent) as chunks are modified, rather than all at once. 

For now in this branch I've created these from scratch each time, but if it turns out that performance is meaningfully impacted when the mempool is busy, then I can optimize this further by just using a bit more memory.

sdaftuar commented on src/policy/rbf.cpp:
```diff
@@ -60,57 +59,26 @@ std::optional<std::string> GetEntriesForConflicts(const CTransaction& tx,
-        for (const CTxIn& txin : mi->GetTx().vin) {
-            parents_of_conflicts.insert(txin.prevout.hash);
+        // Exit early if we're going to fail (see below)
+        if (all_conflicts.size() > MAX_REPLACEMENT_CANDIDATES) {
```
There is room to relax this rule some, so if this is important we can do so. I think the requirement is a bound on the number of clusters that would have to be re-sorted in order to accept the new transaction.  We can approximate that as the number of clusters that would be non-empty as a result of removing all the conflicting transactions from the mempool, and only process replacements for which that is below some target.  

That would be a more complex logic though, so before implementing it I wanted to have some sense of whether we need to.  Has the historical 100-transaction-conflict limit been problematic for use cases in the past?  Note also that in the new code, we are calculating the number of conflicts exactly (the old code used an approximation, which could be gamed by an adversary).

instagibbs replied:
Ah! I wrote a huge response to this, then looked up our previous discussions, and realized I didn't actually read the code: https://github.com/bitcoin/bitcoin/issues/27677#issuecomment-1554708843

IIUC now, this is only counting *direct* conflicts, and not the descendants that are booted.

I think that's fine.

sdaftuar replied:
Ah yes, in the new code I'm only counting direct conflicts right now, because every descendant of a direct conflict must be in the same cluster as that conflict.  So this is already a relaxation of the existing rule.

instagibbs replied:
> I think the requirement is a bound on the number of clusters that would have to be re-sorted in order to accept the new transaction.

As an alternative, we drop the replacement limit to like, 10 or something, and then *only* count the *direct* conflicts, not the direct conflicts and all the descendants?

sdaftuar replied:
I believe I (finally) actually fixed this behavior to count the number of direct conflicts.

sdaftuar replied:
Will comment elsewhere as well, but I just updated this limit to be a limit on the number of distinct clusters that are conflicted.

Sjors commented on src/bench/mempool_stress.cpp:
```diff
@@ -118,5 +199,63 @@ static void MempoolCheck(benchmark::Bench& bench)
+}
+
+BENCHMARK(MemPoolAncestorsDescendants, benchmark::PriorityLevel::HIGH);
+BENCHMARK(MemPoolAddTransactions, benchmark::PriorityLevel::HIGH);
```
dd6684ab665bc8ae76b76fdd2e578fc77d562a52: While you're touching this, can you rename `MempoolCheck` to `MemPoolCheck`, `MempoolEviction` to `MemPoolEviction` and `ComplexMemPool` to `MempoolComplex`? That makes `-filter=MemPool.*` work

As a workaround, `-filter=.*Mem.*` does work.

Sjors commented:
It would be useful to add a `mempool_backwards_compatibility.py` test to illustrate how the new rules interact with older nodes. It could have two modern nodes and one v25 (or v26) node. Some of the tests you deleted in this branch could be moved there. E.g. the test could demonstrate how RBF rule 2 is not enforced when relaying to the new node, but it is when relaying to the v25 node.

Benchmarks on a 2019 MacBook Pro (2,3 GHz 8-Core Intel Core i9), plugged in:

```
% src/bench/bench_bitcoin -filter=.*Mem.* -min-time=10000

|      330,557,188.67 |                3.03 |    1.5% |     10.77 | `ComplexMemPool`
|      451,529,273.50 |                2.21 |    2.8% |     10.01 | `MemPoolAddTransactions`
|            2,847.13 |          351,231.06 |    2.7% |     10.93 | `MemPoolAncestorsDescendants`
|           11,047.90 |           90,514.97 |    2.5% |     10.69 | `MemPoolMiningScoreCheck`
|        4,328,796.04 |              231.01 |    1.1% |     10.99 | `MempoolCheck`
|           36,268.80 |           27,571.91 |    2.9% |     11.17 | `MempoolEviction`
|        9,123,684.25 |              109.60 |    1.4% |     10.74 | `RpcMempool`
```

Sjors commented on src/txmempool.cpp:
```diff
@@ -1245,3 +1536,169 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+
+    for (auto txentry : txs) {
+        m_chunks.emplace_back(txentry.get().GetModifiedFee(), txentry.get().GetTxSize());
+        m_chunks.back().txs.emplace_back(txentry);
```
54f39ca8f101483f5f82707689ca49431d4091e5: So you're creating a chunk for each new transaction and then erasing it if the fee rate goes down. Why not the other way around?

Sjors commented on src/txmempool.cpp:
```diff
@@ -1245,3 +1536,169 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+        while (m_chunks.size() >= 2) {
+            auto cur_iter = std::prev(m_chunks.end());
+            auto prev_iter = std::prev(cur_iter);
+            double feerate_prev = prev_iter->fee*cur_iter->size;
```
54f39ca8f101483f5f82707689ca49431d4091e5: shouldn't this be `fee / size`? Or do you mean to use an inverse fee rate for performance (`inv_fee_rate`)?

Sjors reviewed (commented):
Couple more comments / questions. To be continued...

Sjors commented on src/txmempool.h:
```diff
@@ -632,6 +634,22 @@ class CTxMemPool
 
+    /**
+     * Calculate whether cluster size limits would be exceeded if a new tx were
+     * added to the mempool (assuming no conflicts).
```
065e18fff30a91c94c47d5c2fc65b13ddc38aa47: do you plan to relax this assumption so that a transaction in (a) full cluster(s) can be replaced?

Sjors commented on src/rpc/mempool.cpp:
```diff
@@ -553,6 +660,39 @@ static RPCHelpMan getmempooldescendants()
+    return RPCHelpMan{"getmempoolcluster",
+        "\nReturns mempool data for given cluster\n",
+        {
+            {"id", RPCArg::Type::NUM, RPCArg::Optional::NO, "The cluster id (must be in mempool)"},
```
0b284b5fd29c06d46c1ec60ea7e1bcd07f36feb1: it would be more practical to (also) have a lookup by transaction hash

Sjors commented on src/rpc/mempool.cpp:
```diff
@@ -249,34 +249,81 @@ static RPCHelpMan testmempoolaccept()
+    info.pushKV("txcount", (int)c.m_tx_count);
+    info.pushKV("clusterid", (int)c.m_id);
+    UniValue chunks(UniValue::VARR);
+    for (auto &chunk : c.m_chunks) {
```
0b284b5fd29c06d46c1ec60ea7e1bcd07f36feb1: can you order chunks by mining score?

sdaftuar replied:
Chunks for a given cluster are already sorted in descending feerate order, is that what you're asking about or is there another issue I'm overlooking?

Sjors commented on src/txmempool.cpp:
```diff
@@ -764,48 +918,174 @@ void CTxMemPool::check(const CCoinsViewCache& active_coins_tip, int64_t spendhei
+    int64_t b_size = b->m_cluster->m_chunks[b->m_loc.first].size;
+
+    int64_t a_score = a_fee * b_size;
+    int64_t b_score = b_fee * a_size;
```
It would be good to explain the rationale for the miner score somewhere. Why a * b and b * a?

Sjors commented on src/rpc/mempool.cpp:
```diff
@@ -377,6 +427,63 @@ UniValue MempoolToJSON(const CTxMemPool& pool, bool verbose, bool include_mempoo
+                accum_size += best_chunk.first->size;
+                accum_fee += best_chunk.first->fee;
+
+                UniValue o(UniValue::VOBJ);
```
0b284b5fd29c06d46c1ec60ea7e1bcd07f36feb1: this should also list the cluster id, and maybe also the chunk number.

Could also add an argument limit the total number of vbytes.

Sjors commented on src/rpc/mempool.cpp:
```diff
@@ -377,6 +427,63 @@ UniValue MempoolToJSON(const CTxMemPool& pool, bool verbose, bool include_mempoo
+            }
+            // Define comparison operator on our heap entries (using feerate of chunks).
+            auto cmp = [](const Cluster::HeapEntry& a, const Cluster::HeapEntry& b) {
+                return a.first->fee*b.first->size < b.first->fee*a.first->size;
```
0b284b5fd29c06d46c1ec60ea7e1bcd07f36feb1 : Could this use `CompareMiningScore` or does it have a different goal?

Sjors commented on doc/policy/mempool-replacements.md:
```diff
@@ -53,12 +46,10 @@ other consensus and policy rules, each of the following conditions are met:
    transactions, each with large descendant sets.
 
-6. The replacement transaction's feerate is greater than the feerates of all directly conflicting
+5. The replacement transaction's mining score is greater than the mining score of all directly conflicting
```
bf467f8286425b692a1736ab6d417d0ba6074658 Maybe make this rule 7. That seems a bit more clear than saying "previously this rule referred to fee rate"

Sjors commented on doc/policy/mempool-replacements.md:
```diff
@@ -18,14 +18,7 @@ other consensus and policy rules, each of the following conditions are met:
+2. The replacement transaction pays an absolute fee of at least the sum paid by the original
    transactions.
 
    *Rationale*: Only requiring the replacement transaction to have a higher feerate could allow an
```
"a higher feerate or (using clusters) mining score"

Sjors commented on doc/policy/mempool-replacements.md:
```diff
@@ -53,12 +46,10 @@ other consensus and policy rules, each of the following conditions are met:
+   *Rationale*: Ensure that the new transaction is more appealing to mine than those being evicted.
 
 This set of rules is similar but distinct from BIP125.
 
```
bf467f8286425b692a1736ab6d417d0ba6074658: history can be expanded:

```md
  * Cluster mempool introduced, dropping rule 2 and introducing rule 7. As of **v27.0** ([PR 28676](https://github.com/bitcoin/bitcoin/pull/28676)).
```

Sjors commented on src/policy/rbf.cpp:
```diff
@@ -139,11 +107,11 @@ std::optional<std::string> PaysMoreThanConflicts(const CTxMemPool::setEntries& i
+        Cluster::Chunk &chunk = mi->m_cluster->m_chunks[mi->m_loc.first];
+        CFeeRate original_feerate(chunk.fee, chunk.size);
         if (replacement_feerate <= original_feerate) {
             return strprintf("rejecting replacement %s; new feerate %s <= old feerate %s",
```
bf467f8286425b692a1736ab6d417d0ba6074658: `; new chunk feerate %s <= old chunk feerate`

Sjors commented:
I was wondering if we can drop RBF rule 5 (in a followup), but I'm guessing not really. My initial thinking was the cluster limit could be used instead. But `CalculateMiningScoreOfReplacementTx` only checks that the _new_ cluster doesn't get too big.

But does the new cluster system make it less of a burden to have large numbers of transactions appear and disappear from the mempool?

Sjors commented on src/txmempool.cpp:
```diff
@@ -1245,3 +1648,251 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+template <typename SetType>
+std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef> InvokeSort(size_t tx_count, const std::vector<Cluster::Chunk>& chunks)
+{
+    std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef> txs;
```
Can you clarify what txs, orig_txs and cluster are for, and what the general strategy is here?

Sjors commented:
> I don't think a detailed line-by-line code review at this stage is a good use of your time.

That's not what I'm trying to do. I strategically / haphazerdly picked a dozen lines to ask questions about to get a better understanding of the whole thing. Based on IRC chat today, I'll wait now for the upcoming update.

sdaftuar commented on doc/policy/mempool-replacements.md:
```diff
@@ -18,14 +18,7 @@ other consensus and policy rules, each of the following conditions are met:
-   *Rationale*: When RBF was originally implemented, the mempool did not keep track of
-   ancestor feerates yet. This rule was suggested as a temporary restriction.
-
-3. The replacement transaction pays an absolute fee of at least the sum paid by the original
```
I'm not sure the best way to document things, but I added a to-do item to the OP so that we don't lose track of figuring out how we want the docs to look. (Will have to conform all the code comments as well.)

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -1135,29 +1457,67 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
+            return a.first->fee*b.first->size > b.first->fee*a.first->size;
+        };
+
+        std::make_heap(heap_chunks.begin(), heap_chunks.end(), cmp);
```
Should be fixed now.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -1245,3 +1536,169 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+{
+    m_chunks[entry.m_loc.first].txs.erase(entry.m_loc.second);
+
+    // Chunk (or cluster) may now be empty, but this will get cleaned up
```
I wouldn't say it's safe to ignore, but the idea is that we often want to be able to batch deletions, and then clean things up in one pass.  So the call sites should all be dealing with this issue and ensuring that we always clean up at some point.

(This is definitely an area where I expect that we'll be re-engineering all this logic and trying to come up with a better abstraction layer so that this is more robust and easier to think about!)

ariard commented:
do you have a simulation environment or script like you’ve done for #25717 to observe the computational complexity diff compared to today’s mempool against different chain of transaction performance payload ?
interesting what is the lowest performance linux host assumed for decentralization of the tx-relay network.
assume always 24/7 internet and on a vpcu instance, not baremetal.

Sjors commented:
In the context of Stratum v2 #29432 I'm looking for a more efficient way to decide when to generate a new block template.

The current implementation looks at the mempool's `GetTransactionsUpdated()` every `-sv2interval` seconds. If there was any update, it calls `CreateNewBlock()` on `BlockAssembler`. It then checks if fees increased by at least `-sv2feedelta` since the last template and if so, pushes it to connected clients.

`GetTransactionsUpdated()` changes all the time, especially with a big mempool, so not that useful of a filter. I'd like something like `GetFeeEstimateAtTop(size vbytes=4000000)`. Would it be easy to keep track of that (approximate) number every time something is added or removed from the mempool?

Sjors commented:
Indeed, I meant on top of this PR (or a later incarnation).

ariard commented:
> interesting what is the lowest performance linux host assumed for decentralization of the tx-relay network.
assume always 24/7 internet and on a vpcu instance, not baremetal.

Running this branch mainet on a 2 vcpu instance, with the following performance characteristics:
```
processor       : 0
vendor_id       : AuthenticAMD
cpu family      : 25
model           : 1
model name      : AMD EPYC 7543 32-Core Processor
stepping        : 1
microcode       : 0xa0011d1
cpu MHz         : 2794.750
cache size      : 512 KB
```

It would be very interesting to feed cluster-support testing node with all kinds of chain of transactions, and check substantial performance diff compared to non-cluster-support testing node. Check there is no substantial performance regression.

ariard commented on src/rpc/mempool.cpp:
```diff
@@ -247,34 +247,81 @@ static RPCHelpMan testmempoolaccept()
+    return {
+        RPCResult{RPCResult::Type::NUM, "vsize", "virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted."},
+        RPCResult{RPCResult::Type::NUM, "txcount", "number of transactions (including this one)"},
+        RPCResult{RPCResult::Type::NUM, "clusterid", "id of the cluster containing this tx"},
```
About the `clusterid` (`m_next_cluster_id`), I don’t think it’s a good idea to use it as a tie-breaker in `CompareMiningScore` as it’s just a monotonic counter, it’s not information perfomative for a mining score. You could use weight at least or reception timestamp e.g oldest the tx-relay reception more well-propagated on the network, already in every miner mempool.

ariard commented on src/rpc/mempool.cpp:
```diff
@@ -370,6 +420,76 @@ UniValue MempoolToJSON(const CTxMemPool& pool, bool verbose, bool include_mempoo
+static RPCHelpMan getmempoolfeesize()
+{
+    return RPCHelpMan{"getmempoolfeesize",
+        "Returns fee/size data for the whole mempool.",
```
Obviously you can add fee/weight units here.

ariard commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,157 @@
+ * The size of a FeeFrac cannot be zero unless the fee is also zero.
+ *
+ * FeeFracs have a total ordering, first by increasing feerate (ratio of fee over size), and then
+ * by decreasing size. The empty FeeFrac (fee and size both 0) sorts last. So for example, the
```
This is not perfect as you could use weight units and even have to consider what is the minimal satisfying witness under given consensus rules for a given transaction, especially if you have multiple candidate for a spend given a `witnessScript`. You have already such types of transaction like second-stage LN transactions (either revocation path or preimage/timeout path).

ariard commented on src/util/feefrac.h:
```diff
@@ -0,0 +1,157 @@
+ *
+ * These comparisons are only guaranteed to be correct when the product of the highest fee and
+ * highest size does not exceed 2^64-1. If the fee is a number in sats, and size in bytes, then
+ * this allows up to 46116.86 BTC at size 4M, and 1844674.4 BTC at size 100k).
```
I think this should be `MAX_MONEY` as `ReplacementChecks()` in `PreChecks()` are done before `PolicyScriptChecks()` and `ConsensusScriptChecks()` so an adversary does not have to own the solution of the scriptpubkey to fake `46116.86` or `1844674.4` of value in your local mempool, exceeds 2^64-1 and mess up with linearization.

ariard reviewed (commented):
I’ll spend time reviewing current `CTxMemPool`’s code path w.r.t DoS resistance before to pursue review further. This is very unsure the introduction of `TxGraph` improves on this front as the design is focus on mining incentive-compatibility improvement first.

ariard commented:
> Running this branch mainet on a 2 vcpu instance, with the following performance characteristics:

Got a disk space issue at block `501691` by syncing from genesis.

```
2024-02-23T10:53:41Z UpdateTip: new best=000000000000000000753b6b9821750d271e1730d7403a0658c507c88092bdf0 height=501691 version=0x20000000 log2_work=87.760876 tx=287312998 date='2017-12-30T07:48:59Z' progress=0.295839 cache=16.3MiB(151170txo)
2024-02-23T10:53:41Z New outbound-full-relay v1 peer connected: version: 70016, blocks=831673, peer=196
2024-02-23T10:53:47Z *** Disk space is too low!
2024-02-23T10:53:47Z Error: Disk space is too low!
```

Was configured as a prune node with default pruning settings.
```
2024-02-22T17:49:33Z Prune configured to target 550 MiB on disk for block and undo files.
```

With following `/proc/meminfo` dump

```
MemTotal:        4020688 kB
MemFree:          359104 kB
MemAvailable:    3650432 kB
Buffers:           44424 kB
Cached:          3385624 kB
SwapCached:            0 kB
Active:          1312988 kB
Inactive:        2156904 kB
Active(anon):        352 kB
Inactive(anon):    40004 kB
Active(file):    1312636 kB
Inactive(file):  2116900 kB
```

sdaftuar committed:
Add test case for cluster size limits to TRUC logic

sdaftuar committed:
Eliminate use of cached ancestor data in miniminer_tests and truc_policy

sdaftuar committed:
Use txgraph to calculate ancestors

sdaftuar committed:
Use txgraph to calculate descendants

sdaftuar committed:
Make getting parents/children a function of the mempool, not a mempool entry

sdaftuar committed:
Rewrite GatherClusters to use the txgraph implementation

sdaftuar committed:
Stop tracking parents/children outside of txgraph

sdaftuar committed:
doc: Update mempool_replacements.md to reflect feerate diagram checks

sdaftuar committed:
test: add functional test for new cluster mempool RPCs

sdaftuar committed:
fuzz: remove comparison between mini_miner block construction and miner

This is in preparation for eliminating the block template building happening in
mini_miner, in favor of directly using the linearizations done in the mempool.

Christewart commented on src/kernel/txgraph.cpp:
```diff
@@ -0,0 +1,1352 @@
+    // The first transaction gets to stay in the existing cluster.
+    bool first = true;
+    for (auto& txentry : txs) {
+        if (txentry.get().m_cluster == nullptr) {
```
Isn't this trivially true from this: https://github.com/bitcoin/bitcoin/pull/28676/files#diff-7dcb5aa190f52e5da3aca406127995f65dab3ee9d33ab68130cb3460d2298c7eR498

sdaftuar replied:
See line 528 below: https://github.com/sdaftuar/bitcoin/blob/c49e0444e5c474d5e1ac0af89e9bc27958f3ed31/src/kernel/txgraph.cpp#L528

Christewart commented on src/rpc/mempool.cpp:
```diff
@@ -547,6 +645,40 @@ static RPCHelpMan getmempooldescendants()
+    return RPCHelpMan{"getmempoolcluster",
+        "\nReturns mempool data for given cluster\n",
+        {
+            {"id", RPCArg::Type::NUM, RPCArg::Optional::NO, "The cluster id (must be in mempool)"},
```
How are users of the RPC expected to obtain `cluster_id`? 

IIUC, the way to do this currently is to query 

```
./bin/bitcoin-cli getrawmempool true
```

and then aggregate the `cluster_id`'s yourself?

sdaftuar replied:
Sjors suggested here https://github.com/bitcoin/bitcoin/pull/28676#discussion_r1399137900 that I add a way to do it by transaction hash, which would be easy. Let me know if you have other suggestions...

sdaftuar committed:
Eliminate RBF workaround for CPFP carveout transactions

The new cluster mempool RBF rules take into account clusters sizes exactly, so
with the removal of descendant count enforcement this idea is obsolete.

MarcusMWilliams commented:
Tire & Rim

On Tue, Jun 18, 2024, 3:07 PM DrahtBot ***@***.***> wrote:

> 🚧 At least one of the CI tasks failed. Make sure to run all tests
> locally, according to the
> documentation.
>
> Possibly this is due to a silent merge conflict (the changes in this pull
> request being
> incompatible with the current code in the target branch). If so, make sure
> to rebase on the latest
> commit of the target branch.
>
> Leave a comment here, if you need help tracking down a confusing failure.
>
> Debug: https://github.com/bitcoin/bitcoin/runs/26377000968
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/28676#issuecomment-2176868650>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASHJW52K2WVOETMXWX3XK4TZICHR3AVCNFSM6AAAAAA6F74GWOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNZWHA3DQNRVGA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

instagibbs reviewed (commented):
(after fixing the reported iterations done) Seeing a lot of linerarizations being done that are kind of odd looking repeats with same number of iterations?
```
2024-07-22T21:01:29.226080Z [bench] InvokeSort linearize cluster: 15 txs, 0.0184ms, 120 iter, 153.1ns/iter
2024-07-22T21:01:29.226348Z [bench] InvokeSort linearize cluster: 15 txs, 0.0187ms, 120 iter, 155.7ns/iter
2024-07-22T21:01:29.226633Z [bench] InvokeSort linearize cluster: 15 txs, 0.0180ms, 120 iter, 149.8ns/iter
2024-07-22T21:01:29.226926Z [bench] InvokeSort linearize cluster: 15 txs, 0.0188ms, 120 iter, 156.8ns/iter
```
what's happening here?

sdaftuar commented on src/kernel/txgraph.cpp:
```diff
@@ -0,0 +1,1353 @@
+        orig_linearization.push_back(i);
+    }
+    cluster_linearize::DepGraph dep_graph(cluster);
+    auto result = cluster_linearize::Linearize(dep_graph, iterations, 0, orig_linearization);
```
Fixed this to use `result.second`, which indicates whether the linearization is optimal.

Also added a wip commit to track iterations done, so that the benchmark logging is more useful for now.

sdaftuar commented:
> (after fixing the reported iterations done) Seeing a lot of linerarizations being done that are kind of odd looking repeats with same number of iterations?
> 
> ```
> 2024-07-22T21:01:29.226080Z [bench] InvokeSort linearize cluster: 15 txs, 0.0184ms, 120 iter, 153.1ns/iter
> 2024-07-22T21:01:29.226348Z [bench] InvokeSort linearize cluster: 15 txs, 0.0187ms, 120 iter, 155.7ns/iter
> 2024-07-22T21:01:29.226633Z [bench] InvokeSort linearize cluster: 15 txs, 0.0180ms, 120 iter, 149.8ns/iter
> 2024-07-22T21:01:29.226926Z [bench] InvokeSort linearize cluster: 15 txs, 0.0188ms, 120 iter, 156.8ns/iter
> ```
> 
> what's happening here?

I observed some similar behavior on my node as well.  Without knowing exactly what was happening here: in the current draft of the PR, the number of iterations in an optimal sort is related to the number of topologically valid orderings of the cluster (since we look at all topologically valid subsets each time we try to find the best candidate), so if there are clusters with similar topologies, you might expect to see the same number of iterations needed.

Additionally, in the current draft, successful RBF candidates cause us to relinearize things twice: once to construct the feerate diagram, and once again when the new transaction is added to the mempool.  The intention is to eliminate the  relinearizations from taking place (the design of the TxGraph was done with this in mind), but that is not yet implemented.  (In the example you pasted, I noticed there are 4 linearizations happening at the same millisecond, so I wonder if RBF may also be involved.)

glozow commented on test/functional/feature_rbf.py:
```diff
@@ -60,14 +61,12 @@ def run_test(self):
+        #self.test_too_many_replacements()
 
-        self.log.info("Running test too many replacements...")
-        self.test_too_many_replacements()
-
-        self.log.info("Running test too many replacements using default mempool params...")
-        self.test_too_many_replacements_with_default_mempool_params()
+        #self.log.info("Running test too many replacements using default mempool params...")
+        #self.test_too_many_replacements_with_default_mempool_params()
```
d0576c134a5
I think we only need to adapt 1 of these 2 tests and it should be quite easy to do. Currently, `test_too_many_replacements` requires a high descendant limit because it's 1 parent with 100 children, but we can just add a `self.generate(node, 1)` in between and it can use node 1 (the default one). Then, we can just delete `test_too_many_replacements_with_default_mempool_params`

```
diff --git a/test/functional/feature_rbf.py b/test/functional/feature_rbf.py
index f1de04008c4..57817380c50 100755
--- a/test/functional/feature_rbf.py
+++ b/test/functional/feature_rbf.py
@@ -61,12 +61,8 @@ class ReplaceByFeeTest(BitcoinTestFramework):
         self.log.info("Running test spends of conflicting outputs...")
         self.test_spends_of_conflicting_outputs()
 
-        # TODO: rework too many replacements test to use direct conflicts only
-        #self.log.info("Running test too many replacements...")
-        #self.test_too_many_replacements()
-
-        #self.log.info("Running test too many replacements using default mempool params...")
-        #self.test_too_many_replacements_with_default_mempool_params()
+        self.log.info("Running test too many replacements...")
+        self.test_too_many_replacements()
 
         self.log.info("Running test opt-in...")
         self.test_opt_in()
@@ -331,17 +327,20 @@ class ReplaceByFeeTest(BitcoinTestFramework):
         split_value = int((initial_nValue - fee) / (MAX_REPLACEMENT_LIMIT + 1))
 
         splitting_tx_utxos = self.wallet.send_self_transfer_multi(
-            from_node=self.nodes[0],
+            from_node=self.nodes[1],
             utxos_to_spend=[utxo],
             sequence=0,
             num_outputs=MAX_REPLACEMENT_LIMIT + 1,
             amount_per_output=split_value,
         )["new_utxos"]
 
+        # Let this tx confirm so that its children are not connected
+        self.generate(self.nodes[1], 1)
+
         # Now spend each of those outputs individually
         for utxo in splitting_tx_utxos:
             self.wallet.send_self_transfer(
-                from_node=self.nodes[0],
+                from_node=self.nodes[1],
                 utxo_to_spend=utxo,
                 sequence=0,
                 fee=Decimal(fee) / COIN,
@@ -359,98 +358,13 @@ class ReplaceByFeeTest(BitcoinTestFramework):
         double_tx_hex = double_tx.serialize().hex()
 
         # This will raise an exception
-        assert_raises_rpc_error(-26, "too many potential replacements", self.nodes[0].sendrawtransaction, double_tx_hex, 0)
+        assert_raises_rpc_error(-26, "too many potential replacements", self.nodes[1].sendrawtransaction, double_tx_hex, 0)
 
         # If we remove an input, it should pass
         double_tx.vin.pop()
         double_tx_hex = double_tx.serialize().hex()
         self.nodes[0].sendrawtransaction(double_tx_hex, 0)
 
-    def test_too_many_replacements_with_default_mempool_params(self):
-        """
-        Test rule 5 (do not allow replacements that cause more than 100
-        evictions) without having to rely on non-default mempool parameters.
-
-        In order to do this, create a number of "root" UTXOs, and then hang
-        enough transactions off of each root UTXO to exceed the MAX_REPLACEMENT_LIMIT.
-        Then create a conflicting RBF replacement transaction.
-        """
-        # Clear mempools to avoid cross-node sync failure.
-        for node in self.nodes:
-            self.generate(node, 1)
-        normal_node = self.nodes[1]
-        wallet = MiniWallet(normal_node)
-
-        # This has to be chosen so that the total number of transactions can exceed
-        # MAX_REPLACEMENT_LIMIT without having any one tx graph run into the descendant
-        # limit; 10 works.
-        num_tx_graphs = 10
-
-        # (Number of transactions per graph, rule 5 failure expected)
-        cases = [
-            # Test the base case of evicting fewer than MAX_REPLACEMENT_LIMIT
-            # transactions.
-            ((MAX_REPLACEMENT_LIMIT // num_tx_graphs) - 1, False),
-
-            # Test hitting the rule 5 eviction limit.
-            (MAX_REPLACEMENT_LIMIT // num_tx_graphs, True),
-        ]
-
-        for (txs_per_graph, failure_expected) in cases:
-            self.log.debug(f"txs_per_graph: {txs_per_graph}, failure: {failure_expected}")
-            # "Root" utxos of each txn graph that we will attempt to double-spend with
-            # an RBF replacement.
-            root_utxos = []
-
-            # For each root UTXO, create a package that contains the spend of that
-            # UTXO and `txs_per_graph` children tx.
-            for graph_num in range(num_tx_graphs):
-                root_utxos.append(wallet.get_utxo())
-
-                optin_parent_tx = wallet.send_self_transfer_multi(
-                    from_node=normal_node,
-                    sequence=MAX_BIP125_RBF_SEQUENCE,
-                    utxos_to_spend=[root_utxos[graph_num]],
-                    num_outputs=txs_per_graph,
-                )
-                assert_equal(True, normal_node.getmempoolentry(optin_parent_tx['txid'])['bip125-replaceable'])
-                new_utxos = optin_parent_tx['new_utxos']
-
-                for utxo in new_utxos:
-                    # Create spends for each output from the "root" of this graph.
-                    child_tx = wallet.send_self_transfer(
-                        from_node=normal_node,
-                        utxo_to_spend=utxo,
-                    )
-
-                    assert normal_node.getmempoolentry(child_tx['txid'])
-
-            num_txs_invalidated = len(root_utxos) + (num_tx_graphs * txs_per_graph)
-
-            if failure_expected:
-                assert num_txs_invalidated > MAX_REPLACEMENT_LIMIT
-            else:
-                assert num_txs_invalidated <= MAX_REPLACEMENT_LIMIT
-
-            # Now attempt to submit a tx that double-spends all the root tx inputs, which
-            # would invalidate `num_txs_invalidated` transactions.
-            tx_hex = wallet.create_self_transfer_multi(
-                utxos_to_spend=root_utxos,
-                fee_per_output=10_000_000,  # absurdly high feerate
-            )["hex"]
-
-            if failure_expected:
-                assert_raises_rpc_error(
-                    -26, "too many potential replacements", normal_node.sendrawtransaction, tx_hex, 0)
-            else:
-                txid = normal_node.sendrawtransaction(tx_hex, 0)
-                assert normal_node.getmempoolentry(txid)
-
-        # Clear the mempool once finished, and rescan the other nodes' wallet
-        # to account for the spends we've made on `normal_node`.
-        self.generate(normal_node, 1)
-        self.wallet.rescan_utxos()
-
     def test_opt_in(self):
         """Replacing should only work if orig tx opted in"""
         tx0_outpoint = self.make_utxo(self.nodes[0], int(1.1 * COIN))
```

sdaftuar replied:
Thanks, took this.

glozow reviewed (commented):
had a look at some tests that are disabled or have TODOs

sdaftuar commented on test/functional/mining_prioritisetransaction.py:
```diff
@@ -109,6 +109,14 @@ def test_diamond(self):
         raw_before[txid_c]["fees"]["descendant"] += fee_delta_c_1 + fee_delta_c_2
         raw_before[txid_d]["fees"]["ancestor"] += fee_delta_b + fee_delta_c_1 + fee_delta_c_2
         raw_after = self.nodes[0].getrawmempool(verbose=True)
+        for txid in [txid_a, txid_b, txid_c, txid_d]:
```
Taken, thanks.

sdaftuar commented on test/functional/mempool_persist.py:
```diff
@@ -97,6 +97,9 @@ def run_test(self):
         last_entry = self.nodes[0].getmempoolentry(txid=last_txid)
+        del last_entry["fees"]["chunk"]
+        del last_entry["clusterid"]
+        del last_entry["chunksize"]
```
Taken

sdaftuar commented on test/functional/mempool_persist.py:
```diff
@@ -131,7 +134,11 @@ def run_test(self):
+        new_entry = self.nodes[0].getmempoolentry(txid=last_txid)
+        del new_entry["fees"]["chunk"]
+        del new_entry["clusterid"]
+        del new_entry["chunksize"]
```
Taken

sdaftuar commented on test/functional/mempool_sigoplimit.py:
```diff
@@ -163,14 +163,17 @@ def create_bare_multisig_tx(utxo_to_spend=None):
-        assert "too-long-mempool-chain" in res["tx-results"][tx_child.getwtxid()]["error"]
+        assert "too-large-cluster" in res["tx-results"][tx_child.getwtxid()]["error"]
+        # When we actually try to submit, the parent makes it into the mempool, but the child would exceed cluster vsize limits
+        #assert_raises_rpc_error(-26, "too-large-cluster", self.nodes[0].submitpackage, [tx_parent.serialize().hex(), tx_child.serialize().hex()])
```
Removed

sdaftuar commented on test/functional/mempool_truc.py:
```diff
@@ -557,10 +545,11 @@ def test_truc_sibling_eviction(self):
-        tx_v3_child_2_rule5 = self.wallet.create_self_transfer_multi(utxos_to_spend=utxos_for_conflict, fee_per_output=4000000, version=3)
-        rule5_str = f"too many potential replacements (including sibling eviction), rejecting replacement {tx_v3_child_2_rule5['txid']}; too many potential replacements (101 > 100)"
-        assert_raises_rpc_error(-26, rule5_str, node.sendrawtransaction, tx_v3_child_2_rule5["hex"])
+        # TODO: rewrite this test given the new RBF rules
```
Deleted.

sdaftuar commented:
Thanks @glozow for the test improvements.  I've also made a change to an RBF limit that I think is worth mentioning.  Rather than limit the number of direct conflicts that a transaction can have, I've now implemented the limit to be on the number of distinct clusters that the conflicts of a transaction belong to.

The idea is still to create some bound on the amount of CPU we might spend linearizing clusters and doing feerate diagram checks when processing a single transaction.  Limiting the number of conflicting clusters (at 100 for now) should be sufficient to achieve that.  I want to point out that we might linearize more than 100 clusters as a result of a replacement (because removing conflicts might cause clusters to split), but since linearizing small clusters is generally much less work than linearizing big clusters, I don't think this distinction materially affects the way we understand the implied CPU bound.

instagibbs commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -18,7 +18,7 @@
     def set_test_params(self):
         self.num_nodes = 1
-        self.extra_args = [['-limitdescendantsize=1000', '-limitancestorsize=1000', '-limitancestorcount=100']]
+        self.extra_args = [['-limitdescendantsize=1000', '-limitancestorsize=1000', '-limitancestorcount=100', "-limitclustersize=1000"]]
```
the rest of these args can be deleted and the test passes

sdaftuar replied:
Thanks, removed.

instagibbs reviewed (commented):
poked at the function tests a bit

sdaftuar commented on test/functional/mempool_packages.py:
```diff
@@ -33,8 +29,8 @@ def set_test_params(self):
             [
-                "-limitancestorcount={}".format(CUSTOM_ANCESTOR_LIMIT),
                 "-limitdescendantcount={}".format(CUSTOM_DESCENDANT_LIMIT),
+                "-limitclustercount={}".format(CUSTOM_DESCENDANT_LIMIT),
```
Thanks! Taken.

sdaftuar commented on test/functional/mempool_persist.py:
```diff
@@ -131,7 +132,9 @@ def run_test(self):
-        assert_equal(last_entry, self.nodes[0].getmempoolentry(txid=last_txid))
+        new_entry = self.nodes[0].getmempoolentry(txid=last_txid)
+        del new_entry["clusterid"]
+        assert_equal(last_entry, new_entry)
```
Taken.

sdaftuar commented on test/functional/mempool_truc.py:
```diff
@@ -556,11 +544,6 @@ def test_truc_sibling_eviction(self):
         tx_v3_replacement_only = self.wallet.create_self_transfer_multi(utxos_to_spend=utxos_for_conflict, fee_per_output=4000000)
         # Override maxfeerate - it costs a lot to replace these 100 transactions.
         assert node.testmempoolaccept([tx_v3_replacement_only["hex"]], maxfeerate=0)[0]["allowed"]
-        # Adding another one exceeds the limit.
```
Ah, right.  Given that this falls back to the vsize limit being exercised, I think this isn't really worth including?  Will leave this for now.

sdaftuar committed:
Allow moving an Epoch::Marker

sdaftuar committed:
Allow moving CTxMemPoolEntry objects, disallow copying

sdaftuar committed:
Make CTxMemPoolEntry derive from TxGraph::Ref

sdaftuar committed:
Create a txgraph inside CTxMemPool

sdaftuar committed:
Add transactions to txgraph, but without cluster dependencies

Effectively this is treating all transactions in txgraph as being in a cluster
of size 1.

sdaftuar committed:
Check cluster limits when using -walletrejectlongchains

sdaftuar committed:
Rework truc_policy to use descendants, not children

sdaftuar committed:
Eliminate CheckPackageLimits, which no longer does anything

sdaftuar commented:
Rebased on #31553.

sdaftuar committed:
Add sigops adjusted weight calculator

sdaftuar committed:
Add accessor for sigops-adjusted weight

sdaftuar committed:
Avoid violating mempool policy limits in tests

Changes AddToMempool() helper to only apply changes if the mempool limits are
respected.

Fix package_rbf fuzz target to handle mempool policy violations

sdaftuar committed:
mempool: Store iterators into mapTx in mapNextTx

This takes the same amount of space as CTransaction pointers, and saves a map
lookup in many common uses.

sdaftuar committed:
==== END OPTIMIZATIONS ====

sdaftuar committed:
==== BEGIN DOCS/TESTING ====

sdaftuar committed:
Rework miner_tests to not require large cluster limit

sdaftuar committed:
Fix miniminer_tests to work with cluster limits

sdaftuar committed:
test: update feature_rbf.py replacement test

Preparatory commit to the rbf functional test, before changes are made to the
rbf rules as part of cluster mempool.

instagibbs commented on src/validation.cpp:
```diff
@@ -1073,21 +986,6 @@ bool MemPoolAccept::ReplacementChecks(Workspace& ws)
-    }
 
     CTxMemPool::setEntries all_conflicts;
 
```
Up to this point you have not applied dependencies, so the new transaction is being considered its own singleton cluster, which is in-congruent with the actual goal. If you apply dependencies via CheckMemPoolPolicyLimits, then the resulting cluster is porentially oversized and violates the requirements set forth in the TxGraph interface.

see:
```
diff --git a/src/validation.cpp b/src/validation.cpp
index 61576851cd..affababd10 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -987,10 +987,18 @@ bool MemPoolAccept::ReplacementChecks(Workspace& ws)
 
     CFeeRate newFeeRate(ws.m_modified_fees, ws.m_vsize);
 
     CTxMemPool::setEntries all_conflicts;
 
+    // This triggers dependency addition, but if this returns false,
+    // this means we're not allowed to call CountDistinctClusters
+    // I think we have to stage removals first, then process deps
+    // and only then call GetEntriesForConflicts
+    // Note: Dependencies *still* haven't been processed either,
+    // so each new tx is still considered its own cluster
+    Assume(m_subpackage.m_changeset->CheckMemPoolPolicyLimits());
+
     // Calculate all conflicting entries and enforce Rule #5.
     if (const auto err_string{GetEntriesForConflicts(tx, m_pool, ws.m_iters_conflicting, all_conflicts)}) {
         return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY,
                              strprintf("too many potential replacements%s", ws.m_sibling_eviction ? " (including sibling eviction)" : ""), *err_string);
     }
diff --git a/test/functional/mempool_cluster.py b/test/functional/mempool_cluster.py
index 3da8b477a2..ec7bd62e1a 100755
--- a/test/functional/mempool_cluster.py
+++ b/test/functional/mempool_cluster.py
@@ -2,10 +2,12 @@
 # Copyright (c) 2024 The Bitcoin Core developers
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 """Test cluster mempool accessors and limits"""
 
+from decimal import Decimal
+
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.wallet import (
     MiniWallet,
 )
 from test_framework.util import (
@@ -23,10 +25,11 @@ class MempoolClusterTest(BitcoinTestFramework):
         self.wallet = MiniWallet(node)
 
         node = self.nodes[0]
         parent_tx = self.wallet.send_self_transfer(from_node=node)
         utxo_to_spend = parent_tx["new_utxo"]
+        last_utxo_to_spend = utxo_to_spend
         ancestors = [parent_tx["txid"]]
         while len(node.getrawmempool()) < MAX_CLUSTER_COUNT:
             next_tx = self.wallet.send_self_transfer(from_node=node, utxo_to_spend=utxo_to_spend)
             # Confirm that each transaction is in the same cluster as the first.
             assert node.getmempoolcluster(next_tx['txid']) == node.getmempoolcluster(parent_tx['txid'])
@@ -38,10 +41,11 @@ class MempoolClusterTest(BitcoinTestFramework):
             # Confirm that each successive transaction is added as a descendant.
             assert all([ next_tx["txid"] in node.getmempooldescendants(x) for x in ancestors ])
 
             # Update for next iteration
             ancestors.append(next_tx["txid"])
+            last_utxo_to_spend = utxo_to_spend
             utxo_to_spend = next_tx["new_utxo"]
 
         assert node.getmempoolcluster(parent_tx['txid'])['txcount'] == MAX_CLUSTER_COUNT
         feeratediagram = node.getmempoolfeeratediagram()
         last_val = [0, 0]
@@ -52,10 +56,16 @@ class MempoolClusterTest(BitcoinTestFramework):
 
         # Test that adding one more transaction to the cluster will fail.
         bad_tx = self.wallet.create_self_transfer(utxo_to_spend=utxo_to_spend)
         assert_raises_rpc_error(-26, "too-large-cluster", node.sendrawtransaction, bad_tx["hex"])
 
+        from pdb import set_trace
+        set_trace()
+        # Test that replacing the last tx with RBF works
+        rbf = self.wallet.create_self_transfer(utxo_to_spend=last_utxo_to_spend, fee_rate=Decimal('0.006'))
+        txid = node.sendrawtransaction(rbf["hex"])
+
         # TODO: verify that the size limits are also enforced.
         # TODO: add tests that exercise rbf, package submission, and package
         # rbf and verify that cluster limits are enforced.
 
 if __name__ == '__main__':
```

sdaftuar replied:
I think the mistake here is that in my call to `CountDistinctClusters()`, I should be setting `main_only` to true, and it looks like I omitted that so it'll call it on the staging graph instead.  The same issue may apply with the invocation of `GetDescendants()`; the intent is to just call it on the main graph.

Would that resolve the concern you're bringing up, or is there something else I'm overlooking?

instagibbs replied:
>  I should be setting main_only to true,

I think that works? Will think more on it.

> GetDescendants

Missed that one!

dergoegge commented:
Functional test in this commit: https://github.com/dergoegge/bitcoin/commit/6329ce979f63b396aa036a1ad39798bb83fa4ade (change to `random.cpp` is necessary to make it deterministic, without it the test only hits the assert 1/5 times)

cc @instagibbs

instagibbs commented:
@dergoegge awesome coverage and functional test!

Seems this is the issue, we are `visit`ing children erroneously disallowing their removal later when parent is `visit`ed for immaturity:
```
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index 5c90bc43dd..3c278f6aad 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -362,5 +362,5 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
         WITH_FRESH_EPOCH(m_epoch);
         for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
-            if (!visited(it) && check_final_and_mature(it)) {
+            if (check_final_and_mature(it) && !visited(it)) {
                 txToRemove.emplace_back(it);
                 auto descendants = m_txgraph->GetDescendants(*it);
```

sdaftuar commented:
> I fuzzed this branch ([3bfaedb](https://github.com/bitcoin/bitcoin/commit/3bfaedbd7d9d6e71e9df03fbbb51a33c6184cca6)) with [fuzzamoto](https://github.com/dergoegge/fuzzamoto) and it found an assertion crash in `CTxMemPool::check`.

@dergoegge @instagibbs Thank you both for detecting and pinpointing the bug! Should be fixed now in this branch.

sdaftuar committed:
Invoke TxGraph::DoWork() at appropriate times

sdaftuar commented:
I'm unable to reproduce the ci failure when I run that job locally -- anyone have an idea?

glozow reviewed (commented):
Would it be a good idea to split this PR into a main PR with Cluster Implementation + Cleanups + Docs and Tests, then a followups PR with Optimizations and more cleanups? Goal is to get everything in v31, but this might help with reducing the size of the main PR and make it easier to punt cleanup-related comments to later.

sdaftuar committed:
Update comments for CTxMemPool class

glozow committed:
[test] rework/delete feature_rbf tests requiring large clusters

glozow reviewed (commented):
In the "Still to do" section of the PR description, it seems like most can be checked off other than

> Updating wallet code to be cluster-aware (including mini_miner and coin selection)
> Rework many of our functional tests to be cluster-aware
> Figure out what package validation and package RBF rules should be in this design

sipa reviewed (commented):
Some comments already, mostly to get familiar with the overall flow of changes.

instagibbs reviewed (commented):
few comments through "initial implementation"

instagibbs reviewed (commented):
first pass review through 7683eeed07050fbcfa4f3b77d32303c177e2683e

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -5407,7 +5407,7 @@ class CompareInvMempoolOrder
         /* As std::make_heap produces a max-heap, we want the entries with the
          * fewest ancestors/highest fee to sort later. */
-        return m_mempool->CompareDepthAndScore(*b, *a);
+        return m_mempool->CompareMiningScoreWithTopology(*b, *a);
```
I think this type of p2p change is something we should discuss separately, outside of this PR, because it's not clear to me that there's a simple way to make a strict improvement in relay behavior.  And the interpretation of what the `feefilter` message means/should mean is something we should document and coordinate across implementations as well.

sdaftuar committed:
Use named constant for acceptable iters

sdaftuar commented on src/validation.cpp:
```diff
@@ -1108,6 +1086,12 @@ bool MemPoolAccept::ReplacementChecks(Workspace& ws)
         m_subpackage.m_changeset->StageRemoval(it);
     }
+
+    if (const auto err_string{ImprovesFeerateDiagram(*m_subpackage.m_changeset)}) {
```
Is it possible that we might have failure in the individual setting with a cluster size limit being exceeded, which then is not the case in a package setting, if some additional package transaction conflicts with transactions in the cluster being added to?

sdaftuar commented on src/init.cpp:
```diff
@@ -641,6 +641,9 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
     argsman.AddArg("-test=<option>", "Pass a test-only option. Options include : " + Join(TEST_OPTIONS_DOC, ", ") + ".", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-limitclustercount=<n>", strprintf("Do not accept transactions connected to <n> or more existing in-mempool transactions (default: %u)", DEFAULT_CLUSTER_LIMIT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-limitclustersize=<n>", strprintf("Do not accept transactions whose size with all in-mempool connected transactions exceeds <n> kilobytes (default: %u)", DEFAULT_CLUSTER_SIZE_LIMIT_KVB), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-addrmantest", "Allows to test address relay on localhost", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
Thanks! Guessing this was a rebase error; fixed.

sdaftuar commented on src/test/miniminer_tests.cpp:
```diff
@@ -620,29 +626,29 @@ BOOST_FIXTURE_TEST_CASE(calculate_cluster, TestChain100Setup)
+     * txp0     txp1     txp2    ...  txp31  txp32
      *    \    /    \   /   \            \   /
-     *     txc0     txc1    txc2  ...    txc48
+     *     txc0     txc1    txc2  ...    txc31
```
Fixed, thanks.

sdaftuar commented on src/init.cpp:
```diff
@@ -641,6 +641,9 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
     argsman.AddArg("-limitdescendantcount=<n>", strprintf("Do not accept transactions if any ancestor would have <n> or more in-mempool descendants (default: %u)", DEFAULT_DESCENDANT_LIMIT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-limitdescendantsize=<n>", strprintf("Do not accept transactions if any ancestor would have more than <n> kilobytes of in-mempool descendants (default: %u).", DEFAULT_DESCENDANT_SIZE_LIMIT_KVB), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-test=<option>", "Pass a test-only option. Options include : " + Join(TEST_OPTIONS_DOC, ", ") + ".", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-limitclustercount=<n>", strprintf("Do not accept transactions connected to <n> or more existing in-mempool transactions (default: %u)", DEFAULT_CLUSTER_LIMIT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
Done.

sdaftuar commented on src/txgraph.h:
```diff
@@ -14,7 +14,7 @@
 #define BITCOIN_TXGRAPH_H
 
-static constexpr unsigned MAX_CLUSTER_COUNT_LIMIT{64};
+static constexpr unsigned MAX_CLUSTER_COUNT_LIMIT{201};
```
Thank you!  In the process of looking at this again, I noticed that the `mempool_tests` unit test is failing in some intermediate commits here as well.  Will go back over that as well.

sdaftuar replied:
After reviewing the mempool eviction unit test that was failing in an intermediate commit, I ended up reworking the test altogether, so this should be better now.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -413,6 +413,7 @@ static CTxMemPool::Options&& Flatten(CTxMemPool::Options&& opts, bilingual_str&
 CTxMemPool::CTxMemPool(Options opts, bilingual_str& error)
     : m_opts{Flatten(std::move(opts), error)}
 {
+    m_txgraph = MakeTxGraph(64, 101'000, 10'000);
```
Done.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -1414,9 +1428,58 @@ void CTxMemPool::ChangeSet::StageRemoval(CTxMemPool::txiter it)
+    for (const auto &hash : inputs) {
+        std::optional<txiter> piter = m_pool->GetIter(hash);
+        if (piter) {
+            ret.emplace_back(&(**piter));
```
Sounds good, I added a commit that does this, if it looks right I'll squash into place.

sdaftuar replied:
Squashed this commit.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -1414,9 +1428,58 @@ void CTxMemPool::ChangeSet::StageRemoval(CTxMemPool::txiter it)
+std::vector<const TxGraph::Ref*> CTxMemPool::ChangeSet::CalculateParentsOf(const CTransactionRef& tx)
+{
+    std::vector<const TxGraph::Ref*> ret;
+    std::set<Txid> inputs;
```
Addressed with the inlining of `CalculateParentsOf()`.

sdaftuar commented on src/node/miner.cpp:
```diff
@@ -219,224 +206,90 @@ bool BlockAssembler::TestPackage(uint64_t packageSize, int64_t packageSigOpsCost
-                failedTx.insert(iter->GetSharedTx()->GetHash());
-            }
+        int64_t package_sig_ops = 0;
+        std::vector<const CTxMemPoolEntry*> mempool_txs;
```
Done, and I renamed this variable to `chunk_txs` for clarity.

sdaftuar commented on src/node/miner.cpp:
```diff
@@ -219,224 +206,90 @@ bool BlockAssembler::TestPackage(uint64_t packageSize, int64_t packageSigOpsCost
+            m_mempool->SkipBuilderChunk();
+            // This chunk won't fit, so we let it be removed from the heap and
+            // we'll try the next best.
+            // TODO: try to break up this chunk into smaller chunks for
```
Removed.

sdaftuar commented on src/test/util/setup_common.cpp:
```diff
@@ -535,49 +535,63 @@ CMutableTransaction TestChain100Setup::CreateValidMempoolTransaction(CTransactio
         [](const auto& tx){ return std::make_pair(COutPoint(tx->GetHash(), 0), tx->vout[0].nValue); });
     while (num_transactions > 0 && !unspent_prevouts.empty()) {
-        // The number of inputs and outputs are random, between 1 and 24.
+        // The number of inputs and outputs are random, between 1 and 10.
```
Thanks, fixed the comment.

sdaftuar commented on test/functional/mempool_packages.py:
```diff
@@ -7,8 +7,9 @@
 from test_framework.messages import (
-    DEFAULT_ANCESTOR_LIMIT,
     DEFAULT_DESCENDANT_LIMIT,
+    DEFAULT_DESCENDANT_LIMIT,
```
Oops, eliminated now.

sdaftuar commented on src/validation.cpp:
```diff
@@ -953,11 +953,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     ws.m_iters_conflicting = m_pool.GetIterSet(ws.m_conflicts);
 
     // Calculate in-mempool ancestors, up to a limit.
```
Eliminated the comment.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -377,8 +377,8 @@ void CTxMemPool::removeConflicts(const CTransaction &tx)
             const CTransaction &txConflict = it->second->GetTx();
             if (Assume(txConflict.GetHash() != tx.GetHash()))
             {
+                removeRecursive(it->second, MemPoolRemovalReason::CONFLICT);
```
Looks like a rebase issue (I think there may have been an earlier version of the code where there was a bigger change, and this accidental line reordering was what was left over).  Eliminated this commit.

sdaftuar commented on src/txmempool.h:
```diff
@@ -413,7 +413,7 @@ class CTxMemPool
 
 public:
-    indirectmap<COutPoint, const CTransaction*> mapNextTx GUARDED_BY(cs);
+    indirectmap<COutPoint, txiter> mapNextTx GUARDED_BY(cs);
```
I added a later commit to the CTxMemPool class as a whole that adds some documentation for this variable; do you think that is sufficient or would it be helpful to add more?

sdaftuar commented on src/txmempool.h:
```diff
@@ -510,6 +510,7 @@ class CTxMemPool
      */
     void UpdateTransactionsFromBlock(const std::vector<Txid>& vHashesToUpdate) EXCLUSIVE_LOCKS_REQUIRED(cs, cs_main) LOCKS_EXCLUDED(m_epoch);
 
+public:
```
Thanks, fixed.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -1414,9 +1428,58 @@ void CTxMemPool::ChangeSet::StageRemoval(CTxMemPool::txiter it)
+
+std::vector<const TxGraph::Ref*> CTxMemPool::ChangeSet::CalculateParentsOf(const CTransactionRef& tx)
+{
+    std::vector<const TxGraph::Ref*> ret;
```
This function is gone now.

sdaftuar commented on src/node/miner.cpp:
```diff
@@ -185,30 +184,18 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
-bool BlockAssembler::TestPackage(uint64_t packageSize, int64_t packageSigOpsCost) const
+bool BlockAssembler::TestPackage(FeePerWeight package_feerate, int64_t packageSigOpsCost) const
 {
     // TODO: switch to weight-based accounting for packages instead of vsize-based accounting.
```
Thanks, eliminated.

sdaftuar commented on src/txmempool.h:
```diff
@@ -916,6 +917,24 @@ class CTxMemPool
+
+        auto [chunk_entries, chunk_feerate] = *res;
+        for (TxGraph::Ref* ref : chunk_entries) {
+            entries.emplace_back(dynamic_cast<const CTxMemPoolEntry&>(*ref));
```
Fixed.

sdaftuar commented on src/policy/rbf.cpp:
```diff
@@ -62,59 +62,27 @@ std::optional<std::string> GetEntriesForConflicts(const CTransaction& tx,
-                             MAX_REPLACEMENT_CANDIDATES);
-        }
+    // Rule #5: don't consider replacements that conflict directly with more
+    // than MAX_REPLACEMENT_CANDIDATES distinct clusters. This implies a bound
```
Made some improvements, please let me know how it looks now.

sdaftuar commented on src/policy/rbf.cpp:
```diff
@@ -62,59 +62,27 @@ std::optional<std::string> GetEntriesForConflicts(const CTransaction& tx,
-                             nConflictingCount,
-                             MAX_REPLACEMENT_CANDIDATES);
-        }
+    // Rule #5: don't consider replacements that conflict directly with more
```
Done!

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -1414,9 +1428,58 @@ void CTxMemPool::ChangeSet::StageRemoval(CTxMemPool::txiter it)
+        ProcessDependencies();
+    }
+
+    return !m_pool->m_txgraph->IsOversized();
```
Thanks, fixed.

instagibbs commented on src/validation.cpp:
```diff
@@ -952,12 +952,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-    } else {
-        return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "too-long-mempool-chain", util::ErrorString(ancestors).original);
-    }
+    ws.m_ancestors = m_subpackage.m_changeset->CalculateMemPoolAncestors(ws.m_tx_handle);
```
6000fc8c3fcc0446b3665f7fccf102068919fa3c

Didn't nail down exactly when this would become a concern, but do note that this call can trigger `tx.vin.size()` number staged parents and `GetAncestors` calls based on those parents.

Think this is potentially a huge number of clusters with each up to 64 transactions at a time.

sdaftuar replied:
Good catch -- I'm working on a solution to this, possibly involving a more extensive rewrite of the TRUC checks (to not need parents/ancestors explicitly calculated) and the RBF checks (to not require ancestors to be calculated before we verify that the cluster size limits are respected).

sdaftuar replied:
I added a commit to the end of this PR that should fix this issue.  As I was working on this, I was thinking about your comment above about reworking the TRUC checks to just grab the clusters it needs directly from txgraph.  The issue with that in the way our code works right now is that we can't fully realize our staging clusters until we've added all the transactions, applied dependencies, and staged all the RBF removals.

If we went ahead and imposed a topology requirement on transaction packages so that we know they're staged for addition in a topologically valid order, then we could do a bit better by applying dependencies as we go, so that we could try to look at staging clusters as we go as well.  However I think that is also confusing (in the same way that our existing validation logic is a bit confusing), compared with if we were able to stage all additions and removals up front, and then work on the resulting staging clusters.  Doing that would require a bigger overhaul of the logic in validation.cpp, which I didn't want to tackle just yet.  

But if anyone sees a clean way to do things that is better than what I'm proposing so far, please let me know.

instagibbs replied:
I agree after some thought that txgraph-based checking is too difficult to do, at least for now. I can foresee some weird complications and I'm unsure I could even describe which series of steps would be correct vs our more bespoke approach currently which sidesteps it.

Re-arrangement of checks seems correct, will take a fresh look at the approach soon.

related: Documentation for `PreChecks` needs to be updated as it does not check package(cluster) limits any longer aside from TRUC checks.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -845,7 +827,10 @@ std::vector<CTxMemPool::indexed_transaction_set::const_iterator> CTxMemPool::Get
     }
-    std::sort(iters.begin(), iters.end(), DepthAndScoreComparator());
+    std::sort(iters.begin(), iters.end(), [this](const CTxMemPool::indexed_transaction_set::const_iterator& a, const CTxMemPool::indexed_transaction_set::const_iterator& b) {
+        LOCK(this->cs); // TODO: this is unnecessary, to quiet a compiler warning
```
Took this patch, thanks.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -803,24 +803,20 @@ void CTxMemPool::check(const CCoinsViewCache& active_coins_tip, int64_t spendhei
-    }
-    return counta < countb;
+
+    return m_txgraph->CompareMainOrder(*i.value(), *j.value()) == std::strong_ordering::less ? true : false;
```
Thanks, I'm taking this change.

sdaftuar replied:
Done in 7cebfec8a6197a260e8c21f4637fad5d60163cfa

sdaftuar committed:
Add check that GetSortedScoreWithTopology() agrees with CompareMiningScoreWithTopology()

We use CompareMiningScoreWithTopology() for sorting transaction announcements
during tx relay, and we use GetSortedScoreWithTopology() in
CTxMemPool::check().

sdaftuar commented on src/validation.cpp:
```diff
@@ -606,8 +596,8 @@ class MemPoolAccept
      * If subpackage size == 1, calls AcceptSingleTransaction() with adjusted ATMPArgs to avoid
-     * package policy restrictions like no CPFP carve out (PackageMempoolChecks)
-     * and creates a PackageMempoolAcceptResult wrapping the result.
+     * package policy restrictions like disabled sibling eviction and
```
Happy to clear this comment up to get rid of double-negatives, but just to make sure I've got this right:
 *  AcceptSubPackage() will, in the case of single transaction acceptance, enable sibling eviction, right?
 *  And in package validation, sibling eviction is disabled?

sdaftuar replied:
Should be better now

sdaftuar committed:
Rework RBF and TRUC validation

Calculating mempool ancestors for a new transaction should not be done until
after cluster size limits have been enforced, to limit CPU DoS potential.

Achieve this by reworking TRUC and RBF validation logic:

- TRUC policy enforcement is now done using only mempool parents of
  new transactions, not all mempool ancestors.
- RBF replacement checks are performed earlier (which allows for checking
  cluster size limits earlier, because cluster size checks cannot happen until
  after all conflicts are staged for removal).
- Verifying that a new transaction doesn't conflict with an ancestor now
  happens later, in AcceptSingleTransaction() rather than in PreChecks(). This
  means that the test is not performed at all in AcceptMultipleTransactions(),
  but in package acceptance we already disallow RBF in situations where a
  package transaction has in-mempool parents.

Also to ensure that all RBF validation logic is applied in both the single
transaction and multiple transaction cases, remove the optimization that skips
the PackageMempoolChecks() in the case of a single transaction being validated
in AcceptMultipleTransactions().

instagibbs commented on src/validation.cpp:
```diff
@@ -1340,14 +1362,6 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
-    }
-
     // Check if the transaction would exceed the cluster size limit.
     if (!m_subpackage.m_changeset->CheckMemPoolPolicyLimits()) {
```
as of "Rework RBF and TRUC validation" this check is now redundant I believe

instagibbs commented on src/validation.cpp:
```diff
@@ -1331,6 +1322,37 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
+        ws.m_state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "too-large-cluster", "");
+        return MempoolAcceptResult::Failure(ws.m_state);
+    }
+
```
00c02b42c6199a6c46573304761056082490c7d4

While we're here: Suggestion to make it more clear as to the roles of the various TRUC checks and that everything works once dependencies are attached.

```Suggestion
    // Either sibling eviction was needed, or package based check should always pass for single tx given SingleTRUCChecks already passed
    Assume(ws.m_sibling_eviction || !PackageTRUCChecks(m_pool, ws.m_ptx, ws.m_vsize, {ws.m_ptx}, ws.m_parents));
```

glozow commented on src/validation.cpp:
```diff
@@ -1440,6 +1322,37 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
+    }
+
+    // Check if the transaction would exceed the cluster size limit.
+    if (!m_subpackage.m_changeset->CheckMemPoolPolicyLimits()) {
```
Noting that this check is redundant with `ReplacementChecks`, since `CalculateChunksForRBF` also calls it before calculating the feerate diagram. I think checking twice is fine, but that one has a slightly different error message that should be unified imo. See this test:

```
diff --git a/test/functional/mempool_cluster.py b/test/functional/mempool_cluster.py
index 3da8b477a2f..8bcc9b2fe65 100755
--- a/test/functional/mempool_cluster.py
+++ b/test/functional/mempool_cluster.py
@@ -4,6 +4,11 @@
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 """Test cluster mempool accessors and limits"""
 
+from decimal import Decimal
+
+from test_framework.messages import (
+    COIN,
+)
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.wallet import (
     MiniWallet,
@@ -22,6 +27,7 @@ class MempoolClusterTest(BitcoinTestFramework):
         node = self.nodes[0]
         self.wallet = MiniWallet(node)
 
+        self.log.info("Test that cluster count limit is enforced")
         node = self.nodes[0]
         parent_tx = self.wallet.send_self_transfer(from_node=node)
         utxo_to_spend = parent_tx["new_utxo"]
@@ -54,6 +60,19 @@ class MempoolClusterTest(BitcoinTestFramework):
         bad_tx = self.wallet.create_self_transfer(utxo_to_spend=utxo_to_spend)
         assert_raises_rpc_error(-26, "too-large-cluster", node.sendrawtransaction, bad_tx["hex"])
 
+        self.log.info("Test that cluster count limit is enforced during replacement")
+        utxo_to_double_spend = self.wallet.get_utxo(confirmed_only=True)
+        fee = Decimal("0.000001")
+        tx_to_replace = self.wallet.create_self_transfer(utxo_to_spend=utxo_to_double_spend, fee=fee)
+        node.sendrawtransaction(tx_to_replace["hex"])
+
+        bad_tx_also_replacement = self.wallet.create_self_transfer_multi(
+            utxos_to_spend=[utxo_to_spend, utxo_to_double_spend],
+            fee_per_output=int(fee * COIN * 5),
+        )
+        # Actually, the error is replacement-failed, cluster size limit exceeded (from CalculateChunksForRBF)
+        assert_raises_rpc_error(-26, "too-large-cluster", node.sendrawtransaction, bad_tx_also_replacement["hex"])
+
         # TODO: verify that the size limits are also enforced.
         # TODO: add tests that exercise rbf, package submission, and package
         # rbf and verify that cluster limits are enforced.

```

sdaftuar committed:
bench: rewrite ComplexMemPool to not create oversized clusters

sdaftuar committed:
test: rewrite PopulateMempool to not violate mempool policy (cluster size) limits
